src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -3344,53 +3410,29 @@ def _process_case_value(self, selector, node, parent):
                 new_parent.addchild(leop)
         else:
             # The case value is some scalar expression
-            bop = BinaryOperation(BinaryOperation.Operator.EQ,
-                                  parent=parent)
-            self.process_nodes(parent=bop, nodes=[selector])
-            self.process_nodes(parent=bop, nodes=[node])
-            # TODO #1799 when generic child.datatype is supported we can
-            # remove the conditional inside the loop and support full
-            # expressions
-
-            # Keep track of whether we know if the operator should be EQ/EQV
-            operator_known = False
-            for child in bop.children:
-                if (isinstance(child, Literal) and
-                        isinstance(child.datatype, ScalarType)):
-                    # We know the operator for all literals
-                    operator_known = True
-                    if (child.datatype.intrinsic ==
-                            ScalarType.Intrinsic.BOOLEAN):
-                        rhs = bop.children[1].detach()
-                        lhs = bop.children[0].detach()
+            fake_parent = Assignment(parent=parent)
+            self.process_nodes(parent=fake_parent, nodes=[selector])
+            self.process_nodes(parent=fake_parent, nodes=[node])
+
+            for op in fake_parent.lhs, fake_parent.rhs:
+                if hasattr(op, "datatype") and isinstance(op.datatype, ScalarType):
+                    if op.datatype.intrinsic == ScalarType.Intrinsic.BOOLEAN:
```

Comment:
Did we decide that we would replicate Fortran's weird requirement of having a different comparison operator for bools because it guarantees that we generate correct code in the backend, even when we don't know the type of a Symbol? (Otherwise, I would argue that we should replicate Fortran's weirdness...)

Reply:
Also, worth a comment to explain that Fortran/PSyIR requires EQV when comparing bools.

Reply:
> Did we decide that we would replicate Fortran's weird requirement of having a different comparison operator for bools because it guarantees that we generate correct code in the backend

I don't know if we decided it but we need to because EQ and EQV have different precedent priorities which would require tree modifications in the frontend. Also we wouldn't be able to do the backend properly without knowing the datatype of the expression, which we often don't. So it is much easier to keep both operators.

To problem is when we want to add a new one, e.g. in the select canonicalization, which we can not use EQ or EQV directly, then the solution is using this new interface.

I will explain this is a comment.

Reply:
Great, thanks.